### PR TITLE
join cluster (automatic cluster configuration)

### PR DESCRIPTION
**Pull request by speeedmaster**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Automatic join cluster.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by ianton-ru** [comment]
What about standalone keeper without http(s) port? How to remove keeper host from raft if needed? Is this configuration survives restarts different keeper nodes?

**Comment by antonio2368** [comment]
> What about standalone keeper without http(s) port?

Standalone keeper should have HTTP(s) support. If for some reason you don't want to enable it for your instance (standalone or embedded) supporting the ReconfigRequest from ZK (implemented here https://github.com/ClickHouse/ClickHouse/pull/49450) seems like a good enough option. It should behave the same way but with different interface.

> How to remove keeper host from raft if needed? 

This PR is still being reviewed but removing should be supported by HTTP request.

> Is this configuration survives restarts different keeper nodes?

Yes, because NuRaft stores internally config using Raft protocol also (different type of log). ZK requests are stored as `app_log`, while cluster configuration is stored as `conf`. 

Depending on the type, different behavior is produced by NuRaft internally.

If we store Reconfig requests as ZK nodes, we will have to keep in sync both node and NuRafts internal storage. Replaying logs will also be troublesome because nodes can step down if it detects remove server but it doesn't know there will be a newer log later on which will add it back basically breaking the cluster.

Also, goal of this PR is to use NuRafts internal storage as the only source of truth without specification defined inside the config (unless it's a fresh start of course).  NuRaft already has methods for adding and removing servers so we just need to expose the interface to the clients to be able to call these functions, be it with HTTP or with a ZK request.

@ianton-ru please tell me if you have more questions but also please be patient with this PR, it's a work in progress and we are still polishing details about better cluster management.

File: src/Coordination/KeeperDispatcher.cpp [hunk 1]
```diff
@@ -736,29 +759,174 @@ void KeeperDispatcher::updateConfigurationThread()
...
+            LOG_DEBUG(log, "Configuration change size ({})", diff.size());
+
+        for (auto & change : diff)
+        {
+            bool push_result = update_configuration_queue.push(change);
+            if (!push_result)
+                throw Exception(ErrorCodes::SYSTEM_ERROR, "Cannot push configuration update to queue");
+        }
+
+        snapshot_s3.updateS3Configuration(config, macros);
```
Comment (antonio2368) [inline src/Coordination/KeeperDispatcher.cpp:919]: this shouldn't be under the setting, only the cluster changes

File: src/Coordination/KeeperDispatcher.cpp [hunk 2]
```diff
@@ -736,29 +759,174 @@ void KeeperDispatcher::updateConfigurationThread()
...
+    } else {
+        LOG_INFO(log, "Not initial loading and external updates are allowed. Config update will be ignered");
```
Comment (antonio2368) [inline src/Coordination/KeeperDispatcher.cpp:921]:
```suggestion
        LOG_INFO(log, "Not initial loading and external updates are allowed. Config update will be ignored");
```

File: programs/keeper/Keeper.cpp [hunk 3]
```diff
@@ -530,13 +530,15 @@ try
...
+        [&](ConfigurationPtr config, bool initial_loading)
         {
             if (config->has("keeper_server"))
-                tiny_context->updateKeeperConfiguration(*config);
+                tiny_context->updateKeeperConfiguration(*config, initial_loading);
         },
         /* already_loaded = */ false);  /// Reload it right now (initial loading)
 
+    
+
```
Comment (antonio2368) [inline programs/keeper/Keeper.cpp:541]:
```suggestion
```

**Review by antonio2368 — COMMENTED** [review]
You need to start the http server in `Keeper.cpp` and add the handler so it can be used with standalone Keeper also

**Pull request closed without merging**